packages/orga/src/types.ts

```diff
@@ -181,11 +184,37 @@ export interface Link extends Literal {
   search?: string | number;
 }
 
+export type FootnoteRef = FootnoteReference | FootnoteInline | FootnoteAnonymous;
```

Comment:
I did wonder whether to combine `FootnoteInline` and `FootnoteAnonymous` - I originally split these because of the spec, but they can easily be combined.

It turns out that Org lets you have an _empty_ definition for an inline footnote, so I don't think we could fully combine these (we wouldn't be able to tell if a footnote is inline with an empty definition, or just a reference), but we could do the following:

```typescript
export interface FootnoteReference extends Node {
  type: 'footnote.reference';
  label: string
}

export interface FootnoteInline extends Parent {
  type: 'footnote.inline';
  label?: string; // anonymous when undefined
  children: PhrasingContent[];
}
```

This way we can still easily tell apart anonymous and inline footnotes, and it is probably a good refactoring.

Reply:
Oh hmm, I think this is conflicting with the definition of `Node`, which allows for any `unknown` keys - which means that `label` is always allowed in `FootnoteInline`. So you can always do e.g., `x.label.length`, which can throw an error.

In the interest of type safety, it might make sense for us to keep these separate, as is?

Reply:
Is there any practical difference between a normal footnote reference and an inline one with an empty definition? Anyway, if there is, we can always put an empty literal in the children to signal that it's inline.

Reply:
For the type safety issue, I think it's okay to make `label` non-optional, we consider it's anonymous when length is 0. in fact it's aligned with the spec more literally this way: in `[fn::definition]`, the label technically is an empty string. we can make sense of the definition part the same way. `[fn:label:]` has a definition of an empty string.

Reply:
Re practical difference: yup - one defines a new footnote (albeit an empty one - use cases: maybe they are doing something with the footnote in post), and one references an existing footnote.

Re safety: Ah yep, that's actually a great way to think about it - being an empty string. AFAIK there isn't any way the user can force it to be an empty string, so should be safe.

Reply:
cool, so are we agreeing on merging the type? proposal:
```javascript
export interface FootnoteReference extends Parent {
  type: 'footnote.reference';
  label: string; // anonymous when empty string
  // normal footnote refernece when empty, 
  // could include empty literal (`{type: 'text.plain', value: ''}`) to represents `[fn:label:]`
  children: PhrasingContent[];
}
```

Reply:
@xiaoxinghu I've merged the `FootnoteInline` and `FootnoteAnonymous` types---I'm still not convinced about merging those with the `FootnoteReference` though, for the reasons mentioned above---Org allows for defining _empty_ footnotes, which is different to referencing a footnote. That is to say, `[fn:foo:]` is not the same as `[fn:foo]` - the first defines an inline footnote `foo` with an empty body, which can then be referenced elsewhere; the latter references the existing `foo` footnote.

Reply:
the difference in the syntax tree would be: 
`[fn:foo]` will have an empty array for it's `children`, and `[fn:foo:]` will have children with one node in it, `{type: 'text.plain', value: ''}`. Which aligns with the original org-mode source.

Reply:
Oh I see! Sorry was being slow, yes I agree in this case :)

Reply:
Oh I see! I was being slow - I agree :)